src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,429 @@
+    CLUSTER.createTopic(outputTopic);
+  }
+
+  /**
+   * Implements the stream-side join behavior of waiting a configurable amount of time for table-side data to arrive
+   * before sending a join output for a newly received stream-side record.
+   *
+   * This behavior will increase the likelihood of "fully populated" join output messages, i.e. with data from both the
+   * stream and the table side.  The downside is that this behavior will increase the end-to-end processing latency for
+   * a stream-side record in the topology.
+   */
+  private static final class StreamTableJoinStreamWaitsForTable
+    implements TransformerSupplier<String, Double, KeyValue<String, Pair<Double, Long>>> {
+
+    private final Duration maxWaitTimePerRecordForTableSideData;
+    private final Duration frequencyToCheckForExpiredWaitTimes;
+    private final String streamBufferStoreName;
+    private final String tableStoreName;
+
+    StreamTableJoinStreamWaitsForTable(final Duration maxWaitTimePerRecordForTableData,
+                                       final Duration frequencyToCheckForExpiredWaitTimes,
+                                       final String streamBufferStoreName,
+                                       final String tableStoreName) {
+      this.maxWaitTimePerRecordForTableSideData = maxWaitTimePerRecordForTableData;
+      this.frequencyToCheckForExpiredWaitTimes = frequencyToCheckForExpiredWaitTimes;
+      this.streamBufferStoreName = streamBufferStoreName;
+      this.tableStoreName = tableStoreName;
+    }
+
+    @Override
+    public Transformer<String, Double, KeyValue<String, Pair<Double, Long>>> get() {
+      return new Transformer<String, Double, KeyValue<String, Pair<Double, Long>>>() {
+
+        private KeyValueStore<String, Pair<Double, Instant>> streamBufferStore;
+        private KeyValueStore<String, Long> tableStore;
+        private ProcessorContext context;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+          tableStore = (KeyValueStore<String, Long>) context.getStateStore(tableStoreName);
+          this.context = context;
+          // Note: In practice, you will probably want to use `PunctuationType.STREAM_TIME`.  However, using stream time
+          // would make the test/validation setup in this example more complicated, hence we use wall clock time.
+          this.context.schedule(frequencyToCheckForExpiredWaitTimes, PunctuationType.WALL_CLOCK_TIME, this::punctuate);
+        }
+
+        @Override
+        public KeyValue<String, Pair<Double, Long>> transform(final String key, final Double value) {
+          return sendFullJoinRecordOrWaitForTableSide(key, value);
+        }
+
+        private KeyValue<String, Pair<Double, Long>> sendFullJoinRecordOrWaitForTableSide(final String key, final Double value) {
```

Comment:
Why do we need this method? It's only called once and `transform()` above is a single line. Might be better to embed the method into `transform` directly.

Reply:
Code readability (matter of style/preference).  Rather than introducing code comments to describe what `transform()` does, I rather use a (sub)method with a descriptive name.

Reply:
Subjective -- it's a nit so just leave as is.